src/main/java/org/spongepowered/common/effect/particle/SpongeParticleHelper.java

```diff
@@ -75,6 +85,40 @@ private static int getBlockState(SpongeParticleEffect effect, Optional<BlockStat
         }
     }
 
+    private static int getDirectionData(Direction direction) {
+        if (direction.isSecondaryOrdinal()) {
+            direction = Direction.getClosest(direction.asOffset(), Direction.Division.CARDINAL);
+        }
+        switch (direction) {
+            case SOUTHEAST:
+                return 0;
+            case SOUTH:
+                return 1;
+            case SOUTHWEST:
+                return 2;
+            case EAST:
+                return 3;
+            case WEST:
+                return 5;
+            case NORTHEAST:
+                return 6;
+            case NORTH:
+                return 7;
+            case NORTHWEST:
+                return 8;
+            default:
+                return 4;
+        }
+    }
+
+    private static final EntityFireworkRocket FIREWORK_ROCKET_DUMMY = new EntityFireworkRocket(null);
```

Comment:
The `DESTROY_FIREWORK_ROCKET_DUMMY` and `FIREWORK_ROCKET_DUMMY_EFFECT` are never going to change. And the `FIREWORK_ROCKET_DUMMY` is only used to set the position, id, uniqueId in the packet (This won't cause problems unless you want to use it async). That is also the reason why I didn't use the `EntityDataManager` constructor in `SPacketEntityMetadata`, because the `DataEntry` are shared.